simple_history/utils.py

```diff
@@ -1,11 +1,184 @@
+
+    :param only_for_model: Only disable history creation for this model type.
+    :param instance_predicate: Only disable history creation for model instances passing
+        this predicate.
+    """
+    assert (  # nosec
+        _StoredDisableHistoryInfo.get() is None
+    ), "Nesting 'disable_history()' contexts is undefined behavior"
+
+    if only_for_model:
+        # Raise an error if it's not a history-tracked model
+        get_history_manager_for_model(only_for_model)
+        if instance_predicate:
+            raise ValueError(
+                "'only_for_model' and 'instance_predicate' cannot both be provided"
+            )
+        else:
+
+            def instance_predicate(instance: Model):
+                return isinstance(instance, only_for_model)
+
+    info = _StoredDisableHistoryInfo(instance_predicate)
+    info.set()
+    try:
+        yield
+    finally:
+        info.delete()
+
+
+@dataclass(frozen=True)
+class _StoredDisableHistoryInfo:
+    """
+    Data related to how historical record creation is disabled, stored in
+    ``HistoricalRecords.context`` through the ``disable_history()`` context manager.
+    """
+
+    LOCAL_STORAGE_ATTR_NAME: ClassVar = "disable_history_info"
+
+    instance_predicate: Callable[[Model], bool] = None
+
+    def set(self) -> None:
+        setattr(self._get_storage(), self.LOCAL_STORAGE_ATTR_NAME, self)
+
+    @classmethod
+    def get(cls) -> Optional["_StoredDisableHistoryInfo"]:
+        """
+        A return value of ``None`` means that the ``disable_history()`` context manager
+        is not active.
+        """
+        return getattr(cls._get_storage(), cls.LOCAL_STORAGE_ATTR_NAME, None)
+
+    @classmethod
+    def delete(cls) -> None:
+        delattr(cls._get_storage(), cls.LOCAL_STORAGE_ATTR_NAME)
+
+    @staticmethod
+    def _get_storage() -> Local:
+        from .models import HistoricalRecords  # Avoids circular importing
+
+        return HistoricalRecords.context
+
+
+@dataclass(
+    frozen=True,
+    # DEV: Replace this with just `kw_only=True` when the minimum required
+    #      Python version is 3.10
+    **({"kw_only": True} if sys.version_info >= (3, 10) else {}),
+)
+class DisableHistoryInfo:
+    """
+    Provides info on *how* historical record creation is disabled.
+
+    Create a new instance through ``get()`` for updated info.
+    (The ``__init__()`` method is intended for internal use.)
+    """
+
+    _disabled_globally: bool
+    _instance_predicate: Optional[Callable[[Model], bool]]
+
+    @property
+    def not_disabled(self) -> bool:
+        """
+        A value of ``True`` means that historical record creation is not disabled
+        in any way.
+        If ``False``, check ``disabled_globally`` and ``disabled_for()``.
+        """
+        return not self._disabled_globally and not self._instance_predicate
+
+    @property
+    def disabled_globally(self) -> bool:
```

Comment:
nitpick: There is a bit of a precedence set in the API for a `is_` prefix on these types of boolean property checks.